### PR TITLE
Fixes molotov cocktails being used to craft molotov cocktails

### LINKED ISSUE #20871: Molotov Cocktails can be used as a "drink" in crafting for a Molotov Cocktail
**Problem Description**:
Molotov Cocktails used as a drink in the recipe for a molotov cocktail

**What did you expect to happen**: 
creating a molotov cocktail out of another drink that was on the floor

**What happened instead**:
Instead, a molotov cocktail was created out of the molotov cocktail in my hand

**Why is this bad/What are the consequences:**
makes you waste damp rags, doesn't make sense, and if molotovs are made in self defense then you might die because of it

**Steps to reproduce the problem**:
Have damp rags and bottles of alchohol on the ground. Have molotov cocktail in your hand, try to craft molotov cocktail through craftin g menu

**When did the problem start happening**:
1:05, 4/14/23. I do not know how long it's been around, but this was the first time i noticed it

### PR DESCRIPTION
**Pull request by tsp01**
## What Does This PR Do
Fixes #20871
Molotov cocktails are now blacklisted from being used to craft molotov cocktails.
Now blacklisted items can not be used to craft. For instance you can no longer use molotov cocktails to craft molotov cocktails. Before blacklisted items only prevented selecting the craft option for a given item but could still be used in the crafting if a valid crafting item was in the players reach.

## Why It's Good For The Game
Prevents wasting materials and not getting what you intended to craft.  Things working as intended is good.

## Testing
1. spawn as bartender
2. spawn in 5 damp rags
3. grab 5 bottles of whiskey
4. craft 5 molotov cocktails
5. Profit

## Changelog
:cl:
fix: Fixed being able to use molotov cocktails to craft molotov cocktails
/:cl:

### CODE REVIEW

File: code/modules/crafting/craft.dm [hunk 1]
```diff
@@ -197,6 +197,7 @@
 	var/list/parts_used = list()
 	var/list/reagent_containers_for_deletion = list()
 	var/list/item_stacks_for_deletion = list()
+	var/list/blacklisted = list()
```
Comment (Sirryan2002) [inline code/modules/crafting/craft.dm:200]: can you document this variable?

File: code/modules/crafting/craft.dm [hunk 2]
```diff
@@ -248,6 +249,11 @@
 		else
 			for(var/i in 1 to needed_amount)
 				var/atom/movable/part_atom = locate(thing) in (surroundings - parts_used)
+				while(recipe.blacklist.Find(part_atom.type))
```
Comment (Sirryan2002) [inline code/modules/crafting/craft.dm:252]: can this be turned into a for loop instead?

File: code/modules/crafting/craft.dm (hunk 1 above)
Comment (tsp01) [reply code/modules/crafting/craft.dm]: Done.

File: code/modules/crafting/craft.dm (hunk 2 above)
Comment (tsp01) [reply code/modules/crafting/craft.dm]: and done!

File: code/modules/crafting/craft.dm [hunk 3]
```diff
@@ -248,6 +249,12 @@
 		else
 			for(var/i in 1 to needed_amount)
 				var/atom/movable/part_atom = locate(thing) in (surroundings - parts_used)
+				for(var/j in 1 to length(surroundings - parts_used))
+					if(!recipe.blacklist.Find(part_atom.type))
+						break
+					blacklisted += part_atom
+					part_atom = locate(thing) in (surroundings - parts_used - blacklisted)
```
Comment (farie82) [inline code/modules/crafting/craft.dm:256]:
```suggestion
				var/atom/movable/part_atom
				for(var/atom/movable/candidate as anything in (surroundings - parts_used))
					if(!is_type_in_list(candidate, recipe.blacklist)
						part_atom = candidate
						break
```
It took me a while to understand what the loop was doing.
The suggested way will do the following:
Loop through all surrounding atoms minus the parts that are used. Break when it has found an atom which is not part of the recipe's blacklist. Then use that atom as the part.

This way you don't have to do a lookup each loop to find an appropriate atom to check. Also, this ensures that the last atom checked won't get used regardless of type.

File: code/modules/crafting/craft.dm [hunk 4]
```diff
@@ -197,6 +197,7 @@
 	var/list/parts_used = list()
 	var/list/reagent_containers_for_deletion = list()
 	var/list/item_stacks_for_deletion = list()
+	var/list/blacklisted = list() // list of items not to use from the surroundings based on recipe blacklist
```
Comment (farie82) [inline code/modules/crafting/craft.dm:200]:
```suggestion
```
Won't need it with the below suggestion. Be sure to test it btw just in case.

**Review by farie82 — CHANGES_REQUESTED** [review]

**Comment by tsp01** [comment]
Uh, I got nothing for this one.

**Review by Sirryan2002 — APPROVED** [review]

**Review by farie82 — APPROVED** [review]

**Pull request merged**